homeassistant/components/ezviz/switch.py

```diff
@@ -24,61 +120,75 @@ async def async_setup_entry(
 
-    _attr_device_class = SwitchDeviceClass.SWITCH
+    _attr_has_entity_name = True
 
     def __init__(
-        self, coordinator: EzvizDataUpdateCoordinator, serial: str, switch: str
+        self, coordinator: EzvizDataUpdateCoordinator, serial: str, switch_number: int
     ) -> None:
         """Initialize the switch."""
         super().__init__(coordinator, serial)
-        self._name = switch
-        self._attr_name = f"{self._camera_name} {DeviceSwitchType(switch).name.title()}"
+        self._switch_number = switch_number
         self._attr_unique_id = (
-            f"{serial}_{self._camera_name}.{DeviceSwitchType(switch).name}"
+            f"{serial}_{self._camera_name}.{DeviceSwitchType(switch_number).name}"
         )
-
-    @property
-    def is_on(self) -> bool:
-        """Return the state of the switch."""
-        return self.data["switches"][self._name]
+        self.entity_description = SWITCH_TYPES[switch_number]
+        self._attr_is_on = self.data["switches"][switch_number]
 
     async def async_turn_on(self, **kwargs: Any) -> None:
         """Change a device switch on the camera."""
         try:
             update_ok = await self.hass.async_add_executor_job(
-                self.coordinator.ezviz_client.switch_status, self._serial, self._name, 1
+                self.coordinator.ezviz_client.switch_status,
+                self._serial,
+                self._switch_number,
+                1,
             )
 
         except (HTTPError, PyEzvizError) as err:
-            raise PyEzvizError(f"Failed to turn on switch {self._name}") from err
+            raise HomeAssistantError(f"Failed to turn on switch {self.name}") from err
 
         if update_ok:
-            await self.coordinator.async_request_refresh()
+            self._attr_is_on = True
+            self.async_write_ha_state()
 
     async def async_turn_off(self, **kwargs: Any) -> None:
         """Change a device switch on the camera."""
         try:
             update_ok = await self.hass.async_add_executor_job(
-                self.coordinator.ezviz_client.switch_status, self._serial, self._name, 0
+                self.coordinator.ezviz_client.switch_status,
+                self._serial,
+                self._switch_number,
+                0,
             )
 
         except (HTTPError, PyEzvizError) as err:
-            raise PyEzvizError(f"Failed to turn off switch {self._name}") from err
+            raise HomeAssistantError(f"Failed to turn off switch {self.name}") from err
 
         if update_ok:
-            await self.coordinator.async_request_refresh()
+            self._attr_is_on = False
+            self.async_write_ha_state()
+
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        """Handle updated data from the coordinator."""
+        if self.data["switches"].get(self._switch_number) is None:
+            return
+
+        self._attr_is_on = self.data["switches"][self._switch_number]
```

Comment:
```suggestion
        return self.data["switches"].get(self._switch_number)
```
Should return `None` so it goes unknown if we don't know the state.

Reply:
About 3 - 4 updates over a 24 hour period does not return switch status. The frequency is so low it's more of an irritation than ever noticeable. (EZVIZ servers do this, don't know why)
The idea here is to cache the last state (basically tracking state in memory). Would prevent the odd chance of an automation failing due to an invalid unavailable condition.

There's logic in the update coordinator that would raise when the api is not available. The update coordinator also returns available state so this shouldn't be a problem when the token expires or internet drops for example. Setting the switch will also raise if API is not available.  https://github.com/home-assistant/core/blob/52c7ad130d0c65d7b8eb9764ebff59980037936d/homeassistant/helpers/update_coordinator.py#L455C1-L458C52

Changing the turn_on/turn_off logic to return None when exception is raised might be something I missed. Just need to double check how hass handles this.

Reply:
`unavailable` and `unknown` is not the same thing

Reply:
I humbly apologise for my ignorance! Didn't think this one through.

Reply:
Fixed it. Hope this is acceptable?

Reply:
I don't entirely like that it can't go `unavailable` but I see no good way to deal with it. Not sure if it's only related to these switches or you might have same issues in the other platforms as well.